SIL.Windows.Forms.Keyboarding/Linux/IbusKeyboardSwitchingAdaptor.cs

```diff
@@ -290,49 +290,62 @@ private void HandlePreviewKeyDown(object sender, PreviewKeyDownEventArgs e)
 
 			if (_needIMELocation)
 			{
-				SetImePreeditWindowLocationAndSize(control);
+				SetImePreEditWindowLocationAndSize(control);
 				_needIMELocation = false;
 			}
 
 			var key = e.KeyCode;
-			switch (key)
+			if (key == Keys.Escape)
 			{
-				case Keys.Escape:
-					// These should end a preedit, so wait until that has happened
-					// before allowing the key to be processed.
-					ResetAndWaitForCommit(control);
-					return;
-				case Keys.Up:
-				case Keys.Down:
-				case Keys.Left:
-				case Keys.Right:
-				case Keys.Delete:
-				case Keys.PageUp:
-				case Keys.PageDown:
-				case Keys.Home:
-				case Keys.End:
-				case Keys.Back:
-					if (PassSpecialKeyEventToIbus(control, key, e.Modifiers))
-					{
-						// If IBus handled the key we don't want the control to get it. However,
-						// we can't do this in PreviewKeyDown, so we temporarily subscribe to
-						// KeyDown and suppress the key event there.
-						control.KeyDown += HandleKeyDownAfterIbusHandledKey;
-					}
-					return;
+				// These should end a pre-edit, so wait until that has happened
+				// before allowing the key to be processed.
+				ResetAndWaitForCommit(control);
+				return;
 			}
+
+			if (IsKeyHandledByIbus(key))
+			{
+				if (PassSpecialKeyEventToIbus(control, key, e.Modifiers))
+				{
+					// If IBus handled the key we don't want the control to get it. However,
+					// we can't do this in PreviewKeyDown, so we temporarily subscribe to
+					// KeyDown and suppress the key event there.
+					control.KeyDown += HandleKeyDownAfterIbusHandledKey;
+				}
+				return;
+			}
+
 			// pass function keys onto ibus since they don't appear (on mono at least) as WM_SYSCHAR
 			if (key >= Keys.F1 && key <= Keys.F24)
 				PassSpecialKeyEventToIbus(control, key, e.Modifiers);
 		}
 
 		/// <summary>
-		/// Handles a key down. While a preedit is active we don't want the control to handle
+		/// Handles a key down. While a pre-edit is active we don't want the control to handle
 		/// any of the keys that IBus deals with.
 		/// </summary>
 		private void HandleKeyDownAfterIbusHandledKey(object sender, KeyEventArgs e)
 		{
-			switch (e.KeyCode)
+			if (IsKeyHandledByIbus(e.KeyCode) && sender is Control control)
+			{
+				var eventHandler = GetEventHandlerForControl(control);
+				if (eventHandler != null)
+					e.SuppressKeyPress = eventHandler.IsPreeditActive;
+				control.KeyDown -= HandleKeyDownAfterIbusHandledKey;
+			}
+		}
+
+		/// <summary>
+		/// Gets whether the given key is expected to be handled by IBus when a pre-edit is active.
+		/// </summary>
+		/// <remarks>
+		/// REVIEW: During pre-edit, I assume that IBus handles both basic and modified navigation
+		/// keys like Ctrl+Left. ChatGPT says this is true, but I have not been able to verify this
+		/// since I don't know how to set up a Linux/IBus environment where I could test this.
```

Comment:
To test this on Linux:
- install Ubuntu 24.04 in a VM
- install an IME that uses pre-edit: in a terminal run `sudo apt update && sudo apt install ibus-pinyin`
- open Keyboard settings, click "+ Add Input Source", search for Pinyin by clicking on the three dots below English. If you type the search term "pinyin" you'll get "Other". Clicking that shows "Chinese (Pinyin)" which you then can add.
- close the settings
- in the terminal run `ibus restart` (or reboot)
- the language/keyboard picker in the top bar should now show "Chinese (Pinyin)".